res/res_pjsip/redirect.c

```diff
@@ -0,0 +1,403 @@
+/*
+ * Asterisk -- An open source telephony toolkit.
+ *
+ * Copyright (C) 2025, Commend International
+ *
+ * Maximilian Fridrich <m.fridrich@commend.com>
+ *
+ * See http://www.asterisk.org for more information about
+ * the Asterisk project. Please do not directly contact
+ * any of the maintainers of this project for assistance;
+ * the project provides a web site, mailing lists and IRC
+ * channels for your use.
+ *
+ * This program is free software, distributed under the terms of
+ * the GNU General Public License Version 2. See the LICENSE file
+ * at the top of the source tree.
+ */
+
+#include "asterisk.h"
+
+#include <pjsip.h>
+
+#include "asterisk/linkedlists.h"
+#include "asterisk/res_pjsip.h"
+#include "asterisk/res_pjsip_redirect.h"
+
+/*!
+ * \internal
+ * \brief Visited URI tracking for redirect loop detection
+ */
+struct visited_uri {
+	char uri[AST_SIP_MAX_URI_SIZE];
+	AST_LIST_ENTRY(visited_uri) list;
+};
+
+/*!
+ * \internal
+ * \brief Redirect contact with q-value for prioritization
+ */
+struct redirect_contact {
+	char uri[PJSIP_MAX_URL_SIZE];
+	float q_value; /* q-value from Contact header, default 1.0 if not present */
+	AST_LIST_ENTRY(redirect_contact) list;
+};
+
+/*! \brief List of redirect contacts */
+AST_LIST_HEAD_NOLOCK(redirect_contact_list, redirect_contact);
+
+/*!
+ * \brief Redirect context structure
+ */
+struct ast_sip_redirect_context {
+	struct ast_sip_endpoint *endpoint;
+	int hop_count;
+	AST_LIST_HEAD_NOLOCK(, visited_uri) visited_uris;
+	struct redirect_contact_list pending_contacts;
+};
+
+struct ast_sip_redirect_context *ast_sip_redirect_context_create(
+	struct ast_sip_endpoint *endpoint,
+	const char *initial_uri)
+{
+	struct ast_sip_redirect_context *context;
+	struct visited_uri *visited;
+
+	context = ast_calloc(1, sizeof(*context));
+	if (!context) {
+		return NULL;
+	}
+
+	context->endpoint = ao2_bump(endpoint);
+	context->hop_count = 0;
+	AST_LIST_HEAD_INIT_NOLOCK(&context->visited_uris);
+	AST_LIST_HEAD_INIT_NOLOCK(&context->pending_contacts);
+
+	/* Add the initial URI to visited list */
+	if (initial_uri) {
+		visited = ast_calloc(1, sizeof(*visited));
+		if (visited) {
+			ast_copy_string(visited->uri, initial_uri, sizeof(visited->uri));
+			AST_LIST_INSERT_HEAD(&context->visited_uris, visited, list);
+		} else {
+			ast_log(LOG_WARNING, "Failed to allocate initial visited URI - redirect loop detection may be impaired\n");
+		}
+	}
+
+	return context;
+}
+
+int ast_sip_should_redirect(struct ast_sip_endpoint *endpoint, int status_code)
+{
+	/* Check if it's a 3xx response */
+	if (!PJSIP_IS_STATUS_IN_CLASS(status_code, 300)) {
+		return 0;
+	}
+
+	/* Check redirect_method on endpoint */
+	if (endpoint->redirect_method != AST_SIP_REDIRECT_URI_PJSIP) {
+		ast_log(LOG_NOTICE, "Received %d redirect, but redirect_method is not 'uri_pjsip' (it is '%s'). Not following redirect.\n",
```

Comment:
This needs to include the endpoint name. "Received %d redirect while sending MESSAGE to %s"